// @google-cloud/cloud-sql-connector/dist/mjs/cloud-sql-instance.js
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
import { selectIpAddress } from './ip-addresses.js';
import { parseInstanceConnectionName } from './parse-instance-connection-name.js';
import { generateKeys } from './crypto.js';
import { getRefreshInterval } from './time.js';
export class CloudSQLInstance {
    static async getCloudSQLInstance(options) {
        const instance = new CloudSQLInstance(options);
        await instance.refresh();
        return instance;
    }
    constructor({ ipType, authType, instanceConnectionName, sqlAdminFetcher, }) {
        this.closed = false;
        this.ipType = ipType;
        this.authType = authType;
        this.instanceInfo = parseInstanceConnectionName(instanceConnectionName);
        this.sqlAdminFetcher = sqlAdminFetcher;
    }
    async refresh() {
        const rsaKeys = await generateKeys();
        const metadata = await this.sqlAdminFetcher.getInstanceMetadata(this.instanceInfo);
        this.ephemeralCert = await this.sqlAdminFetcher.getEphemeralCertificate(this.instanceInfo, rsaKeys.publicKey, this.authType);
        this.host = selectIpAddress(metadata.ipAddresses, this.ipType);
        this.privateKey = rsaKeys.privateKey;
        this.serverCaCert = metadata.serverCaCert;
        if (!this.closed) {
            this.refreshTimeoutID = setTimeout(() => {
                this.refresh();
            }, getRefreshInterval(this.ephemeralCert.expirationTime));
        }
    }
    close() {
        if (this.refreshTimeoutID) {
            clearTimeout(this.refreshTimeoutID);
        }
        this.closed = true;
    }
}
