// @google-cloud/cloud-sql-connector/dist/cjs/cloud-sql-instance.js
"use strict";
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
Object.defineProperty(exports, "__esModule", { value: true });
exports.CloudSQLInstance = void 0;
const ip_addresses_1 = require("./ip-addresses");
const parse_instance_connection_name_1 = require("./parse-instance-connection-name");
const crypto_1 = require("./crypto");
const time_1 = require("./time");
class CloudSQLInstance {
    static async getCloudSQLInstance(options) {
        const instance = new CloudSQLInstance(options);
        await instance.refresh();
        return instance;
    }
    constructor({ ipType, authType, instanceConnectionName, sqlAdminFetcher, }) {
        this.closed = false;
        this.ipType = ipType;
        this.authType = authType;
        this.instanceInfo = (0, parse_instance_connection_name_1.parseInstanceConnectionName)(instanceConnectionName);
        this.sqlAdminFetcher = sqlAdminFetcher;
    }
    async refresh() {
        const rsaKeys = await (0, crypto_1.generateKeys)();
        const metadata = await this.sqlAdminFetcher.getInstanceMetadata(this.instanceInfo);
        this.ephemeralCert = await this.sqlAdminFetcher.getEphemeralCertificate(this.instanceInfo, rsaKeys.publicKey, this.authType);
        this.host = (0, ip_addresses_1.selectIpAddress)(metadata.ipAddresses, this.ipType);
        this.privateKey = rsaKeys.privateKey;
        this.serverCaCert = metadata.serverCaCert;
        if (!this.closed) {
            this.refreshTimeoutID = setTimeout(() => {
                this.refresh();
            }, (0, time_1.getRefreshInterval)(this.ephemeralCert.expirationTime));
        }
    }
    close() {
        if (this.refreshTimeoutID) {
            clearTimeout(this.refreshTimeoutID);
        }
        this.closed = true;
    }
}
exports.CloudSQLInstance = CloudSQLInstance;
